### PR TITLE
feat: `hide-get-premium` patch

### PR DESCRIPTION
**Pull request by ponces**
Before (original):

After partial patch (hidden bottom panel):

After successful patching:

### CODE REVIEW

File: src/main/kotlin/app/revanced/patches/music/layout/premium/fingerprints/HideGetPremiumFingerprint.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import app.revanced.patcher.fingerprint.method.annotation.MatchingMethod
+import app.revanced.patches.music.layout.premium.annotations.HideGetPremiumCompatibility
+import org.jf.dexlib2.AccessFlags
+import org.jf.dexlib2.Opcode
+
+@Name("hide-get-premium-fingerprint")
+@MatchingMethod(
+    "Lktn;", "k"
+)
+@FuzzyPatternScanMethod(2) // FIXME: Test this threshold and find the best value.
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/music/layout/premium/fingerprints/HideGetPremiumFingerprint.kt:17]:
```suggestion

```

File: src/main/kotlin/app/revanced/patches/music/layout/premium/fingerprints/HideGetPremiumParentFingerprint.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import app.revanced.patcher.fingerprint.method.annotation.MatchingMethod
+import app.revanced.patches.music.layout.premium.annotations.HideGetPremiumCompatibility
+import org.jf.dexlib2.AccessFlags
+import org.jf.dexlib2.Opcode
+
+@Name("hide-get-premium-parent-fingerprint")
+@MatchingMethod(
+    "Lktn;", "i"
+)
+@FuzzyPatternScanMethod(2) // FIXME: Test this threshold and find the best value.
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/music/layout/premium/fingerprints/HideGetPremiumParentFingerprint.kt:17]:
```suggestion

```

File: src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        val parentResult = HideGetPremiumParentFingerprint.result!!
+        HideGetPremiumFingerprint.resolve(data, parentResult.classDef)
+
+        val method = HideGetPremiumFingerprint.result!!.mutableMethod
+
+        val firstIndex = method.implementation!!.instructions.indexOfFirst {
+            ((it as? BuilderInstruction35c)?.reference.toString() == "Landroid/text/TextUtils;->isEmpty(Ljava/lang/CharSequence;)Z")
+        }
+        method.replaceInstruction(
+            firstIndex - 6, """
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt:39]: Is -6 not a bit too far? In the fingerprint `HideGetPremiumFingerprint` add a 1-x opcodes which match to the instructions you need to change.

File: src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            const/4 v1, 0x0
+        """
+        )
+
+        val secondIndex = method.implementation!!.instructions.indexOfFirst {
+            ((it as? BuilderInstruction35c)?.reference.toString() == "Landroid/view/View;->setVisibility(I)V")
+        }
+        method.addInstructions(
+            secondIndex - 4, """
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt:48]: Same here

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt (hunk 3 above)
Comment (ponces) [reply src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt:39]: I'm not using opcodes on the fingerprint, I'm using a string... how can I achieve what you're asking?

Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt:39]: Check how other fingerprints are made with both an opcode pattern and strings

Comment (ponces) [reply src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt:39]: I have to get the index of two places in this (quite huge) method so I'll split into two fingerprint classes to get each index. Is this the right thing to do?

Comment (TheJeterLP) [reply src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt:39]: Theres multiple ways to do what you want to do

Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt:39]: You'd need to simply create a fingerprint to find the method, then a second one with just the opcode pattern and resolve it to the method. The second fingerprint will then contain the information for the scan result.

Comment (ponces) [reply src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt:39]: Ok. But then I should have a third fingerprint with another opcode pattern to get the second place in the method which I want to retrieve. Does this make sense?

Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt:39]: No, that would be done by the first fingerprint already.

Comment (ponces) [reply src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt:39]: Ok, so the first fingerprint should find the method and the first index which I want to change. The second fingerprint should find the second index alone, using the first fingerprint to resolve/find the method. Is this ok now?

Comment (ponces) [reply src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt]: It's working with https://github.com/revanced/revanced-patches/pull/195/commits/f0f5450ea8080069b68fb61041018fd0dd89abfe! Thanks for the help!
Always great to learn from you both :)

File: src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt (hunk 4 above)
Comment (ponces) [reply src/main/kotlin/app/revanced/patches/music/layout/premium/patch/HideGetPremiumPatch.kt]: Fixed on https://github.com/revanced/revanced-patches/pull/195/commits/f0f5450ea8080069b68fb61041018fd0dd89abfe

**Comment by TheJeterLP** [comment]
Looks good to me 👍🏻

**Comment by TheJeterLP** [comment]
I don't know how the target method looks like, but instead of creating two fingerprints which do exactly the same (dupe code) you can make one fingerprint, put all the opcodes in there from your first instruction you want to change until the last instruction you want to change and just do startIndex and endIndex.

**Comment by ponces** [comment]
> I don't know how the target method looks like, but instead of creating two fingerprints which do exactly the same (dupe code) you can make one fingerprint, put all the opcodes in there from your first instruction you want to change until the last instruction you want to change and just do startIndex and endIndex.

Can't do that as the target method is quite huge and the opcodes list would have a length too big. Which in the future can lead to patching failures easily...

**Comment by TheJeterLP** [comment]
> > I don't know how the target method looks like, but instead of creating two fingerprints which do exactly the same (dupe code) you can make one fingerprint, put all the opcodes in there from your first instruction you want to change until the last instruction you want to change and just do startIndex and endIndex.
> 
> Can't do that as the target method is quite huge and the opcodes list would have a length too big. Which in the future can lead to patching failures easily...

Yeah okay then that is not an option 👍🏻

**Pull request merged**